### PR TITLE
add time_zone to FreeBusyView

### PR DESCRIPTION
**Pull request by AssassiaN**
With this change we can get timezone.
In my solution with correct bias but "incorrect" name.
We need timezone to correctly interpret working hours.
This isnt great solution, but i didnt found other, except to build correlation between timezones (which is very time-consuming).

```
accounts = [(account, 'Organizer', False)]
busy_info = account.protocol.get_free_busy_info(accounts=accounts, start=UTC.localize(EWSDateTime.now()), end=UTC.localize(EWSDateTime.now() + timedelta(hours=6)))
busy_info = list(busy_info)[0]

time_zone = busy_info.time_zone
bias = time_zone.bias

now_time = EWSDateTime.now()
if time_zone.daylight_time.iso_month and time_zone.standard_time.iso_month:
	daylight_time = EWSDateTime(now_time.year, time_zone.daylight_time.iso_month,
								time_zone.daylight_time.weekday)
	standard_time = EWSDateTime(now_time.year, time_zone.standard_time.iso_month,
								time_zone.standard_time.weekday)

	if daylight_time > standard_time:
		if standard_time < now_time < daylight_time:
			bias += time_zone.daylight_time.bias
		else:
			bias += time_zone.standard_time.bias
	else:
		if daylight_time < now_time < standard_time:
			bias += time_zone.daylight_time.bias
		else:
			bias += time_zone.standard_time.bias

# bias = abc(bias) // 60
bias = bias // 60

if bias > 0:
	zone_name = 'Etc/GMT+{}'.format(bias)
elif bias < 0:
	zone_name = 'Etc/GMT{}'.format(bias)
else:
	zone_name = 'Etc/GMT'
timezone = EWSTimeZone.timezone(zone_name)

#tests
test_time = EWSDateTime(2018, 9, 28, 12)
tz_kiev = EWSTimeZone.timezone('Europe/Kiev')
print tz_kiev
print tz_kiev.localize(test_time)
print UTC.normalize(tz_kiev.localize(test_time))
print timezone
print timezone.localize(test_time)
print UTC.normalize(timezone.localize(test_time))
```

Same timezone (Europe/Kiev = GMT+3) set in account settings.
Output:
```
Europe/Kiev
2018-09-28 12:00:00+03:00
2018-09-28 09:00:00+00:00
Etc/GMT-3
2018-09-28 12:00:00+03:00
2018-09-28 09:00:00+00:00
```

Only one issue: in time_zone.daylight_time and time_zone.daylight_time 'time' is None.

### CODE REVIEW

**Comment by ecederstrand** [comment]
Working hours are supposed to be in local time. Essentially, no matter where in the world you happen to work in a given week, your working hours will be e.g. Monday to Friday from 8 to 17 *in local time*. If you are working in Singapore, then you meet at the office at 8 AM local time. If you are working in London, which is 7 hours behind Singapore, then you still show up at 8 AM local time, not at 01:00 AM.

If your client code needs to convert working hours to local time because you know where the person works, then just `tz.localize()` the `start` and `end` values of the `WorkingPeriod`.

To sum up, I think it's wrong to make working hours timezone-aware.

**Comment by AssassiaN** [comment]
I didnt want to make working hours timezone-aware. 
We need way to know what timezone is set in account settings.
Are there any other options where to get the timezone, except i have described?
We writing meeting scheduler and we need to compare working hours from peoples with different timezones.
Ex: 2 people with timezones (GMT+2 and GMT+4), both starts working at 8 am, but in UTC 1st starts working at 6am and 2nd at 4am. We need to put the meeting in working time for both.
So we need timezone.

**Comment by ecederstrand** [comment]
I think the best option here is to use TimeZone object to make an educated guess on the proper timezone name by comparing with the output of `account.protocol.get_timezones()`. We would fetch all the timezones the server supports, select the ones where the timezone periods match, and then narrow down the list by further looking at the timezone transitions.

Here's how to list all properties of all server timezones:

```python
from pprint import pprint
for tz_id, tz_name, tz_periods, tz_transitions, tz_transitions_groups in a.protocol.get_timezones(return_full_timezone_data=True):
    pprint(dict(tz_id=tz_id, tz_name=tz_name, tz_periods=tz_periods, tz_transitions=tz_transitions, tz_transitions_groups=tz_transitions_groups))
```

There is already a `TimeZone.from_server_timezone()` helper, and we could create at `TimeZone.to_server_timezone()` helper that does the reverse, as described above. We just need add caching, so we don't generate a full timezone request for every TimeZone object we want to convert.

**Comment by ecederstrand** [comment]
Hi @AssassiaN 

I just pushed a series of commits to fix this issue. I slightly renamed your new 'timezone' field to 'working_hours_timezone' to clearly mark what the timezone relates to. I also added a function to translate from a TimeZone instance to a MS timezone ID, which you can use to translate back to an EWSTimeZone instance:

```python
timezones = list(a.protocol.get_timezones(return_full_timezone_data=True))
accounts = [(a, 'Organizer', False)]
tz = EWSTimeZone.timezone('Europe/Copenhagen')
start = tz.localize(EWSDateTime.now())
end = start + timedelta(hours=6)
for busy_info in a.protocol.get_free_busy_info(accounts=accounts, start=start, end=end):
    print(busy_info.working_hours_timezone)
    ms_id = busy_info.working_hours_timezone.to_server_timezone(timezones, start.year)
    print(ms_id)
    print(EWSTimeZone.from_ms_id(ms_id))
```

Would that work for you?

**Comment by AssassiaN** [comment]
HI
semiworking, not all timezones detects correctly, Europe/Kiev detects as Europe/Bucharest
and crash with exception for others:
"No server timezones match this timezone definition"

About exception, i think here is a problem:
https://github.com/ecederstrand/exchangelib/blob/991699e0d0aa08f648a8c5003ff0226eaab262ef/exchangelib/properties.py#L518
some timezones was return with "daylight_time" and "standard_time" = None
and next line comparison didnt work correctly
100% not working timezones: UTC+6 Astana, UTC+5 Islamabad, UTC+3 Nairobi

**Comment by ecederstrand** [comment]
There is usually never be a 100% match because the TimeZone element returned by EWS in the WorkingHours element simply does not contain enough information for a 100% accurate match for each of the accounts returned.

Probably Europe/Kiev and Europe/Bucharest are in fact the same timezone. I get the same for Europe/Copenhagen, which returns as Europe/Zurich, but it doesn't matter because they are both CET/CEST. Calculation between timezones will still work.

If you have suggestions how to translate the failing timezones correctly, I'm all ears.

**Comment by ecederstrand** [comment]
I'm closing this PR since the main problem is solved. If you want to work on fixing the Islamabad and Nairobi timezone issues, feel free to open a new issue and we can look at the solutions for supporting those.
